### DISCUSSION TITLE
Rust lsp isn't working

### DISCUSSION THREAD

**Discussion by bajrangCoder** [mkslanc/ace-linters#69] [Q&A] [answered]
I have checked [https://mkslanc.github.io/ace-linters/](https://mkslanc.github.io/ace-linters/) but it doesn't works.
is it bug or any other problem. I wanted to use rust lsp in [Acode](https://github.com/deadlyjack/Acode) through web worker.

And also not working for php

**Comment by mkslanc** [comment, marked as answer]
It's working. It could take a lot of time to initialize but after it's working - 
![image](https://github.com/mkslanc/ace-linters/assets/32402726/75660456-4eab-4dc7-ad5a-d8b616b84f3b)

Though, Rust added only as an example how to build your own service based on wasm (for example). It's not supported service for ace linters

**Reply by mkslanc** [reply]
The PHP service only supports validation. I couldn't find any PHP open-source web linter that supports all LSP features, such as hover, completions, and more. I hope I have answered all your questions. If not, feel free to ask more.

**Reply by bajrangCoder** [reply]
ok, one more question: how can i built my own service for any language?

**Reply by mkslanc** [reply]
1. You need to implement the [LanguageService](https://github.com/mkslanc/ace-linters/blob/main/packages/ace-linters/types/language-service.d.ts#L9) interface. You can refer to the [CSS service](https://github.com/mkslanc/ace-linters/blob/main/packages/ace-linters/services/css/css-service.ts#L11C64-L11C64) as an example.

2. You can also utilize the [BaseService](https://github.com/mkslanc/ace-linters/blob/main/packages/ace-linters/services/base-service.ts#L6C23-L6C34) to avoid describing routine operations like `setValue`, `addDocument`, etc.

3. After describing the service, bundle it with any bundler to create a single endpoint.

4. Register your service in the `ServiceManager`, similar to [any of those services](https://github.com/mkslanc/ace-linters/blob/main/packages/demo/webworker-lsp/webworker.ts#L4).

5. Use your web worker with your services in application.

6. Profit!

**Reply by bajrangCoder** [reply]
Ok thanks for your answer
